### PR TITLE
[FLINK-4075] ContinuousFileProcessingCheckpointITCase failed on Travis

### PR DESCRIPTION
**Pull request by kl0u**
This fixes 2 unstable tests:
1) the  ContinuousFileProcessingCheckpointITCase  and
2) the JMXReportedTest.

@zentol please review the fix of the second.

### CODE REVIEW

**Comment by aljoscha** [comment]
I think having `splitsToFwdOrderedAscByModTime` and `currentSplitsToFwd` as fields that are checkpointed is no longer necessary since `monitorDirAndForwardSplits()` is called in lock scope and those two fields are always set to `null` at the end of the method.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 1]
```diff
@@ -104,7 +104,13 @@ public void open() throws Exception {
 		this.collector = new TimestampedCollector<>(output);
 		this.checkpointLock = getContainingTask().getCheckpointLock();
 
+		Preconditions.checkArgument(reader == null, "The reader is already initialized.");
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: I think this should be `Preconditions.checkState()`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 2]
```diff
@@ -104,7 +104,13 @@ public void open() throws Exception {
...
+
 		this.reader = new SplitReader<>(format, serializer, collector, checkpointLock, readerState);
+
+		// after initializing the reader, set the state to recovered state
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: Could you please clarify this sentence. I think something might be missing.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 3]
```diff
@@ -405,6 +414,9 @@ public void restoreState(StreamTaskState state, long recoveryTimestamp) throws E
 		S formatState = (S) ois.readObject();
 
 		// set the whole reader state for the open() to find.
+		Preconditions.checkArgument(this.readerState == null,
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: This should also be `Preconditions.checkState`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 4]
```diff
@@ -191,11 +197,11 @@ public void close() throws Exception {
 
 		private volatile boolean isSplitOpen = false;
 
-		SplitReader(FileInputFormat<OT> format,
+		private SplitReader(FileInputFormat<OT> format,
 					TypeSerializer<OT> serializer,
 					TimestampedCollector<OT> collector,
 					Object checkpointLock,
-					Tuple3<List<FileInputSplit>, FileInputSplit, S> restoredState) {
+					Tuple3<List<FileInputSplit>, FileInputSplit, S> recoveredState) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: I think the old name was fine. In other parts of Flink it's also called restored state.

**Comment by aljoscha** [comment]
Thanks for fixing this! I had some comments but they should be easy to address.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 5]
```diff
@@ -334,9 +342,11 @@ public void run() {
...
-			if (this.format instanceof CheckpointableInputFormat && this.isSplitOpen) {
-				S formatState = (S) ((CheckpointableInputFormat) format).getCurrentState();
-				return new Tuple3<>(snapshot, currentSplit, currentSplit == null ? null : formatState);
+			if (this.format instanceof CheckpointableInputFormat && this.currentSplit != null) {
+				S formatState = this.isSplitOpen ?
+					(S) ((CheckpointableInputFormat) format).getCurrentState() :
+					restoredFormatState;
+				return new Tuple3<>(snapshot, currentSplit, formatState);
 			} else {
 				LOG.info("The format used is not checkpointable. The current input split will be restarted upon recovery.");
```
Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: I think this log message is wrong for `this.currentSplit != null`. It leads to pollution of log files.

Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: Monitoring an empty directory looks like this:

```
11:36:42,704 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Triggering checkpoint 1 @ 1467279402703
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,709 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,710 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:42,823 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Completed checkpoint 1 (in 120 ms)
11:36:44,703 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Triggering checkpoint 2 @ 1467279404703
11:36:44,705 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,706 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,706 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,706 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,706 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,706 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,707 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,710 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:44,745 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Completed checkpoint 2 (in 18 ms)
11:36:46,703 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Triggering checkpoint 3 @ 1467279406703
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,705 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:46,723 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Completed checkpoint 3 (in 8 ms)
11:36:48,703 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Triggering checkpoint 4 @ 1467279408703
11:36:48,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,704 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,705 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,705 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,705 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,705 INFO  org.apache.flink.streaming.api.functions.source.ContinuousFileReaderOperator  - The format used is not checkpointable. The current input split will be restarted upon recovery.
11:36:48,729 INFO  org.apache.flink.runtime.checkpoint.CheckpointCoordinator     - Completed checkpoint 4 (in 10 ms)
```

**Pull request closed without merging**